REFRAME.md

```diff
@@ -0,0 +1,290 @@
+}
+```
+
+Note: Each Request type has a corresponding Response type.
+Every message except the Error type should end in Request/Response.
+
+#### Error
+
+The Error message type should be used in Responses to indicate that an error has occurred.
+
+```ipldsch
+    type Error struct {
+        Method String
+    }
+```
+
+#### Identify
+
+A message for discovering which messages a server supports. May be used by applications to optimize which servers get sent which types of requests.
+
+```ipldsch
+    type IdentifyRequest struct {}
+
+    type IdentifyResponse struct {
+        Methods [String]
+    }
+```
+
+`IdentifyResponse` should return the set of supported methods aside from the "Identify" method.
+
+##### DAG-JSON Examples
+
+Request:
+```
+{"IdentifyRequest" : {}}
+```
+
+Response:
+```
+{"IdentifyResponse" : {
+    "Methods" : ["FindProviders", "GetIPNS"]
+}}
+```
+
+#### FindProviders
+
+A message for finding nodes that have an interest in a given key. Some common examples include finding which peers have advertised that they have a given CID, or which peers are interested in a given pubsub topic.
+
+```ipldsch
+    type FindProvidersRequest struct 
+        Key &Any
+    }
+```
+
+Note: While the Key is a CID it is highly recommended that server implementations treat these FindProvidersRequests as if they were for the multihash. This allows consumers that perform lookups by CID but with a different IPLD codec than one that is known to the server to still get back valid responses.
+
+```ipldsch
+    type FindProvidersResponse struct {
+        Providers [Provider]
+    }
+
+    type Provider struct {
+     Node Node
+     Proto optional [TransferProtocol]
+     }
+
+    # Node describes a node (identity, network address, authentication, whatever else applies)
+    # We expect different types of nodes, e.g. peer, miner, public IP, etc.
+    type Node union {
+     | Peer "peer"
+    } representation keyed
+
+    type Peer struct {
+        ID Bytes # The binary representation of a libp2p peerID https://github.com/libp2p/specs/blob/f433ad595224cf33d916c166d1738f11aadfa9f7/peer-ids/peer-ids.md
+        Multiaddresses optional [Bytes] # Each element in the list is the binary representation of a complete multiaddr without a peerID suffix
+                                        # https://github.com/multiformats/multiaddr/blob/d02c681171629912dd977f4f5d4b3b72dcac507f/README.md
+                                        # https://github.com/libp2p/specs/blob/98e0ca0e54fd83f7b8c86d7e9677e1e430da5810/addressing/README.md#multiaddr-in-libp2p
+    }
+    
+    # An alternative mechanism might be to have a union keyed by the protocol identifier that has more detailed
+    # information inside the bytes. 
+
+    # Note: This is not quite valid IPLD Schema because the "structured" represenation does not exist yet, nor does the SingletonInt mechanism
+    # However, we can use this to describe that we have a union keyed by numbers rather than strings.

```

Comment:
Some alternative representations could have been:

A: 
```
type TransferProtocol struct {
    ID Bytes
    Hint optional Bytes # Or Any?
}
```

and make more of the information about what fields are needed by different protocols live in documentation/not in the schema.

B: 

One of the existing union representations using the integer codes as strings although that seems pretty weird given it's not a string

```
type TransferProtocol union {
     | Bitswap "2304"
     | GraphSync-FILV1 "2320"
} representation keyed
```

C: Use an existing union representation but using protocol names rather than numbers

I don't feel super strongly about the way we encode this and was mostly trying to take requests from other parties into account here. If there are strong preferences to go with one of the above approaches or something else here's the place to make them known.

Reply:
B or C both seem fine to me.  A is viable but sad.

As a general rule, the impetus for choosing B or C is that they mean it's perfectly clear how things end up serialized in DAG-JSON (where we're limited to string keys in maps, by the aim of compatibility with JSON itself).

Beyond that, I tend not to care.  Season to taste.  How much entropy / specificity / descriptivism / whatever-you-want-to-call-it one wants to stir into strings is up to the author.  Stringified magic numbers don't offend me any more than other magic strings.

Reply:
Approach A gives the most open ended holder for totally unknown unstructured data, but at the cost of being the least useful.  E.g. there would almost certainly be a second layer of parsing and pattern matching going on inside that.  Doable but probably will be perceived by most developers as unpleasant and complicated.

Reply:
Discussion had during IPFS standup: we're going to do option B